src/tiered/tiered_work.c

```diff
@@ -9,25 +9,44 @@
 #include "wt_internal.h"
 
 /*
- * __wt_tiered_work_free --
- *     Free a work unit and account for it in the flush state.
+ * __tiered_flush_state --
+ *     Account for flush work units so threads can know when shared storage flushing is complete.
  */
-void
-__wt_tiered_work_free(WT_SESSION_IMPL *session, WT_TIERED_WORK_UNIT *entry)
+static void
+__tiered_flush_state(WT_SESSION_IMPL *session, uint32_t type, bool incr)
 {
     WT_CONNECTION_IMPL *conn;
     uint32_t new_state, old_state;
 
+    if (type != WT_TIERED_WORK_FLUSH)
+        return;
     conn = S2C(session);
     for (;;) {
         WT_BARRIER();
         old_state = conn->flush_state;
-        new_state = old_state - 1;
+        if (incr)
+            new_state = old_state + 1;
+        else {
+            WT_ASSERT(session, old_state != 0);
+            new_state = old_state - 1;
+        }
         if (__wt_atomic_casv32(&conn->flush_state, old_state, new_state))
             break;
         WT_STAT_CONN_INCR(session, flush_state_races);
         __wt_yield();
     }
```

Comment:
Rewrite this to use `wt_atomic_add32()` and `wt_atomic_sub32()`? It would avoid looping around the CAS instruction.

Reply:
Thanks for that fresh perspective! That is so obvious when you point it out. It simplifies so just a couple lines now.